src/Middleware/Diagnostics/src/ExceptionHandler/ExceptionHandlerExtensions.cs

```diff
@@ -95,7 +97,54 @@ public static IApplicationBuilder UseExceptionHandler(this IApplicationBuilder a
                 throw new ArgumentNullException(nameof(options));
             }
 
-            return app.UseMiddleware<ExceptionHandlerMiddleware>(Options.Create(options));
+            var iOptions = Options.Create(options);
+            return SetExceptionHandlerMiddleware(app, iOptions);
+        }
+
+        private static IApplicationBuilder SetExceptionHandlerMiddleware(IApplicationBuilder app, IOptions<ExceptionHandlerOptions>? options)
+        {
+            const string globalRouteBuilderKey = "__GlobalEndpointRouteBuilder";
+            // Check if UseRouting() has been called so we know if it's safe to call UseRouting()
+            // otherwise we might call UseRouting() when AddRouting() hasn't been called which would fail
+            if (app.Properties.TryGetValue("__EndpointRouteBuilder", out _) || app.Properties.TryGetValue(globalRouteBuilderKey, out _))
```

Comment:
We should only do this if we're using `WebApplicationBuilder` so let's nuke this other check. That means you can also just capture the routeBuilder here once.
```suggestion
            if (app.Properties.TryGetValue(globalRouteBuilderKey, out var routeBuilder))
```